### ISSUE TITLE
Find and execute VSTests

### ISSUE THREAD

**Issue by fwinkelbauer** [cake-contrib/Cake.Recipe#50]
Cake.Recipe seems to not find VSTests in Cake.VsMetrics, see conversation [here](https://github.com/fwinkelbauer/Cake.VsMetrics/issues/1).

**Comment by gep13** [comment]
Ok, so this one is a little trickier to handle, but shouldn't be impossible 😄 

I normally use xUnit in my projects, so again, this is another area that isn't particularly well tested.  Let me ask a couple questions..

When you enable VSTest on a project, what project references do you include, and do you know what the differences are between MSTest and VSTest?  Or are they the same, and they just use different test runners?

**Comment by fwinkelbauer** [comment]
> what project references do you include

The testing reference for VSTest is `Microsoft.VisualStudio.QualityTools.UnitTestFramework.dll`. All the setup is done by Visual Studio if I create a new test project. There could be more stuff that goes on under the hood, but that's all I can see in Visual Studio without inspecting the csproj file in detail.

> what the differences are between MSTest and VSTest

From what I now (and what you can find on [StackOverflow](http://sqa.stackexchange.com/questions/7561/how-to-choose-between-tcm-exe-mstest-exe-and-vstest-console-exe#7565)), VSTest is the "new MSTest". So any test project I have I run with VSTest. Checking out xUnit and NUnit is on my TODO list, but for now I take the lazy route 😆

**Comment by gep13** [comment]
@fwinkelbauer from your investigations, did you happen to find out what the project reference was for MSTest?

**Comment by fwinkelbauer** [comment]
I am afraid I don't know, I have only ever really worked with VSTest. I might be able to test some stuff out tomorrow, I'll give you an update if I stumble over something

**Comment by gep13** [comment]
@fwinkelbauer perfect.  I will do some digging as well.

**Comment by fwinkelbauer** [comment]
This is what I found out today: [Microsoft Social](https://social.msdn.microsoft.com/Forums/en-US/54cf8b08-ded9-4420-8bec-225bc0e98984/differences-between-mstest-and-vstest-and-how-to-validate-code-against-specific-version-of-the-net?forum=vsunittest), [Microsoft Blog](https://blogs.msdn.microsoft.com/bhuvaneshwari/2012/06/15/vstest-console-exe-commandline-test-runner/)

According to "Microsoft Blog" VSTest is used since VS2011 in VS itself and VSTest could even be used to run tests from other frameworks. The link also lists a few differences in terms of the available command line parameters.

> [VSTest] won’t save the results and there is no .trx file generated in the target directory.

This quote is from the other link and I believe that it is incorrect. VSTest has a command line parameter which you can use to create a .trx result file. The only downside is that you cannot choose a target location, the .trx file is always created in a TestResults folder in the working directory. I have done this [before](https://github.com/fwinkelbauer/BuildPipeline.Cake) (in this [file](https://github.com/fwinkelbauer/BuildPipeline.Cake/blob/master/SampleApplication/pipeline.cake)).

All this leads us to the big questions: How should we know when to call which test runner (xUnit, NUnit, MSTest, VSTest)? I guess that you are currently searching for some test assemblies to identify test projects (hence the question above, I haven't taken a look at the implementation). The following list are just some brainstormy thoughts on solutions:
- Always use VSTest (if the claim above is right that it can handle other projects)
- Use the assembly to identify which test command to use. MSTest could potentially be using the same reference as VSTest. I don't have a version of Visual Studio that is 2011 or below, so I can't test it
- Let the user decide which test command to use (e.g. using an enum). Test projects could be identified through their name using wildcards.

Does this help?

**Comment by gep13** [comment]
@fwinkelbauer okay, sorry for taking so long to get back to you on this, our second daughter was born on 17th February, so only now getting back on top of things.

I think I have fixed this issue with this commit: https://github.com/cake-contrib/Cake.Recipe/commit/5f85c34806b3e8239911fff877c42ad53436b4ec

From what I can ascertain, the vstest.console.exe is a replacement for the older mstest.exe.  It does somethings that the older exe doesn't, but the older exe is still used for some specific project types.  I have updated Cake.Recipe to default to running vstest.console.exe, and thanks to this [PR](https://github.com/cake-contrib/Cake.Incubator/pull/13) Cake.Incubator is now capable of parsing the conditional reference, which was causing a problem as well.

I am away to push a slight change into the Cake.VsMetrics repo just to verify that everything is working correctly.

Please let me know if you have any questions.

**Comment by fwinkelbauer** [comment]
@gep13 No worries, a friend of mine hangs out in the Choco gitter a lot, so I knew that you have a lot on your plate right now!

I have pulled all the changes and I can confirm that everything works as expected. VSTest is running. That is sweet! It's pretty awesome that conditional references work now in Cake - that's a problem I ran into in [as well](https://github.com/fwinkelbauer/Cake.BuildUtil).
